File: 0733. Flood Fill/733_bfs.py
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, newColor: int) -> List[List[int]]:
        row, col = len(image), len(image[0])
        move = [[-1, 0], [0, 1], [1, 0], [0, -1]]
        visit, queue = [[0]*col for _ in range(row)], [(sr, sc)]
        initColor = image[sr][sc]
        while queue:
            r, c = queue.pop()
            if (0 <= r < row and 0 <= c < col) and image[r][c] == initColor and not visit[r][c]:
                visit[r][c] = 1
                image[r][c] = newColor
                for m in move:
                    tr, tc = r + m[0], c + m[1]
                    queue = [(tr, tc)] + queue

        return image

File: 0733. Flood Fill/733_bfs.cpp
class Solution {
public:
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int color) {
        if(image[sr][sc]==color) return image;

        queue<pair<int, int>> que;
        int m = image.size(), n = image[0].size();
        int oldColor = image[sr][sc];

        que.push({sr, sc});
        while(!que.empty()){
            auto pts = que.front();
            que.pop();

            int r = pts.first, c = pts.second;
            image[r][c] = color;
            if(c-1>=0 && image[r][c-1]==oldColor)
                que.push({r, c-1});
            if(c+1<n && image[r][c+1]==oldColor)
                que.push({r, c+1});
            if(r-1>=0 && image[r-1][c]==oldColor)
                que.push({r-1, c});
            if(r+1<m && image[r+1][c]==oldColor)
                que.push({r+1, c});
        }
        return image;
    }
};
